fastseq/optimizer/transformers/beam_search_optimizer.py

```diff
@@ -115,915 +185,1009 @@ def _update_beam_size(self, num_beams):
-
-            repetition_penalty: (`optional`) float
-                The parameter for repetition penalty. Between 1.0 and infinity.
-                1.0 means no penalty. Default to 1.0.
-
-            pad_token_id: (`optional`) int
-                Padding token. Default to specicic model pad_token_id or None if
-                it does not exist.
-
-            bos_token_id: (`optional`) int
-                BOS token. Defaults to `bos_token_id` as defined in the models
-                config.
-
-            eos_token_id: (`optional`) int
-                EOS token. Defaults to `eos_token_id` as defined in the models
-                config.
-
-            length_penalty: (`optional`) float
-                Exponential penalty to the length. Default to 1.
-
-            no_repeat_ngram_size: (`optional`) int
-                If set to int > 0, all ngrams of size `no_repeat_ngram_size` can
-                only occur once.
-            bad_words_ids: (`optional`) list of lists of int
-                `bad_words_ids` contains tokens that are not allowed to be
-                generated. In order to get the tokens of the words that should
-                not appear in the generated text, use
-                `tokenizer.encode(bad_word, add_prefix_space=True)`.
-
-            num_return_sequences: (`optional`) int
-                The number of independently computed returned sequences for each
-                element in the batch. Default to 1.
-
-            attention_mask (`optional`) obj: `torch.LongTensor` of same shape as
-                `input_ids`. Mask to avoid performing attention on padding token
-                indices. Mask values selected in ``[0, 1]``: ``1`` for tokens
-                that are NOT MASKED, ``0`` for MASKED tokens. Defaults to `None`
-
-                `What are attention masks? <../glossary.html#attention-mask>`__
-
-            decoder_start_token_id=None: (`optional`) int
-                If an encoder-decoder model starts decoding with a different
-                token than BOS. Defaults to `None` and is changed to `BOS`
-                later.
-
-            use_cache: (`optional`) bool
-                If `use_cache` is True, past key values are used to speed up
-                decoding if applicable to model. Defaults to `True`.
-
-            model_specific_kwargs: (`optional`) dict
-                Additional model specific kwargs will be forwarded to the
-                `forward` function of the model.
+        # determine generation mode
+        is_greedy_gen_mode = (num_beams == 1) and (num_beam_groups == 1) and do_sample is False
+        is_sample_gen_mode = (num_beams == 1) and (num_beam_groups == 1) and do_sample is True
+        is_beam_gen_mode = (num_beams > 1) and (num_beam_groups == 1) and do_sample is False
+        is_beam_sample_gen_mode = (num_beams > 1) and (num_beam_groups == 1) and do_sample is True
+        is_group_beam_gen_mode = (num_beams > 1) and (num_beam_groups > 1)
```

Comment:
Are our changes compatible with all these generation modes?

Reply:
The beam search updates are only applied for is_beam_gen_mode. The model-specific updates (ie attention) are applied in all cases.

Reply:
Does it also work with `is_group_beam_gen_mode`?

Reply:
no is_group_beam_gen_mode does not use the updates. Should that be added?

Reply:
If the optimizations can also work for `group_beam_gen_mode`, we can add it. We can support it in the another PR later if it will  take some time to support it.

Reply:
I think it will take some time to make it work for both, so I think it's best if group_beam_gen_mode goes in another pr

Reply:
Sounds good.